test/__snapshots__/test/test.ts.md

```diff
@@ -225147,7 +225124,7 @@ Generated by [AVA](https://avajs.dev).
     /**␊
      * The list of custom alert time-window rules.␊
      */␊
-    timeWindowRules?: ((ActiveConnectionsNotInAllowedRange | AmqpC2DMessagesNotInAllowedRange | MqttC2DMessagesNotInAllowedRange | HttpC2DMessagesNotInAllowedRange | AmqpC2DRejectedMessagesNotInAllowedRange | MqttC2DRejectedMessagesNotInAllowedRange | HttpC2DRejectedMessagesNotInAllowedRange | AmqpD2CMessagesNotInAllowedRange | MqttD2CMessagesNotInAllowedRange | HttpD2CMessagesNotInAllowedRange | DirectMethodInvokesNotInAllowedRange | FailedLocalLoginsNotInAllowedRange | FileUploadsNotInAllowedRange | QueuePurgesNotInAllowedRange | TwinUpdatesNotInAllowedRange | UnauthorizedOperationsNotInAllowedRange)[] | string)␊
+    timeWindowRules?: (TimeWindowCustomAlertRule[] | string)␊
```

Comment:
This looks like a regression. Asked about it here to understand the expected behavior: https://github.com/json-schema-org/json-schema-spec/issues/1529

Reply:
Given this schema:

```yaml
type: object
required: [ruleType, timeWindowSize]

oneOf:
  - $ref: '#/definitions/ActiveConnectionsNotInAllowedRange'
  # ...etc...
  - $ref: '#/definitions/UnauthorizedOperationsNotInAllowedRange'

properties:
  ruleType:
    type: string
    enum: [ TimeWindowCustomAlertRule ]

  timeWindowSize:
    type: string
```

The old behavior is:

```ts
export interface DeviceSecurityGroupProperties {
  // Note: does **not** include the ruleType or timeWindowSize properties
  timeWindowRules?:
    | (
        | ActiveConnectionsNotInAllowedRange
        // ...etc...
        | UnauthorizedOperationsNotInAllowedRange
      )[]
    | string;
  [k: string]: unknown;
}

export type TimeWindowCustomAlertRule = {
  ruleType: "TimeWindowCustomAlertRule";
  timeWindowSize: string;
  [k: string]: unknown;
} & (
  | ActiveConnectionsNotInAllowedRange
  // ...etc...
  | UnauthorizedOperationsNotInAllowedRange
);
```

The new behavior is:

```ts
export interface DeviceSecurityGroupProperties {
  // Note: Now includes ruleType and timeWindowSize.
  timeWindowRules?: TimeWindowCustomAlertRule[] | string;
}

export type TimeWindowCustomAlertRule = {
  ruleType: "TimeWindowCustomAlertRule";
  timeWindowSize: string;
  [k: string]: unknown;
} & (
  | ActiveConnectionsNotInAllowedRange
  // ...etc...
  | UnauthorizedOperationsNotInAllowedRange
);
```

I would argue that the new behavior is more nearly correct. My reasoning is based on the definition of the `oneOf` keyword in the [JSON Schema spec][1]:

> 10.2.1.3. oneOf
>
> This keyword's value MUST be a non-empty array. Each item of the array MUST be a valid JSON Schema.
>
> An instance validates successfully against this keyword if it validates successfully against exactly one schema defined by this keyword's value.

This definition does not address the possibility that `oneOf` may be combined with additional keywords, such as `properties`. However, one reasonable interpretation is that the `properties` keyword should be applied in addition to the `oneOf` keyword. In such a case, both `oneOf` and `properties` (or other keywords) would all apply.

This interpretation aligns with [section 10.1][3]:

> 10.1 Keyword Independence
>
> Schema keywords typically operate independently, without affecting each other's outcomes.

This suggests that `oneOf` and `properties` should _both_ affect the output.

This interpretation is also in line with the [Factoring Schemas][2] example, which demonstrates how one schema:

```yaml
oneOf:
  - { type: number, multipleOf: 5 }
  - { type: number, multipleOf: 3 }
```

Can be factored into a schema that combines `oneOf` with additional keywords (`type`, in this case, but I presume an example with `properties` would be equally valid):

```yaml
type: number
oneOf:
  - { multipleOf: 5 }
  - { multipleOf: 3 }
```

I realize that other interpretations of the `oneOf` schema are possible, but this is what feels most intuitive to me.  Please let me know if other changes need to be made.

[1]: https://json-schema.org/draft/2020-12/json-schema-core#section-10.2.1.3 'oneOf'
[2]: https://json-schema.org/understanding-json-schema/reference/combining#factoringschemas 'Factoring Schemas'
[3]: https://json-schema.org/draft/2020-12/json-schema-core#name-keyword-independence 'Keyword Independence'

Reply:
I think that's a reasonable interpretation, but as you said, there may be other interpretations. Implicit intersection schemas like this make me scratch my head sometimes. I *think* your interpretation is correct, and the new behavior is correct, but since it's a subtle breaking change that will affect the largest consumers of JSTT, I want to double check with JSON-Schema maintainers first.

I bumped the Github issue to see is someone is willing to clarify.